torch_em/loss/distance_based.py

```diff
@@ -0,0 +1,63 @@
+import torch.nn as nn
+
+from .dice import DiceLoss
+
+
+class DistanceLoss(nn.Module):
+    """Loss for distance based instance segmentation.
+
+    Expects input and targets with three channels: foreground and two distance channels.
+    Typically the distance channels are centroid and inverted boundary distance.
+
+    Args:
+        mask_distances_in_bg: whether to mask the loss for distance predictions in the background.
+        foreground_loss: the loss for comparing foreground predictions and target.
+        distance_loss: the loss for comparing distance predictions and target.
+    """
+    def __init__(
+        self,
+        mask_distances_in_bg: bool,
+        foreground_loss: nn.Module = DiceLoss(),
+        distance_loss: nn.Module = nn.MSELoss(reduction="mean")
+    ) -> None:
+        super().__init__()
+
+        self.foreground_loss = foreground_loss
+        self.distance_loss = distance_loss
+        self.mask_distances_in_bg = mask_distances_in_bg
+
+    def forward(self, input_, target):
+        assert input_.shape == target.shape, input_.shape
+        assert input_.shape[1] == 3, input_.shape
+
+        # IMPORTANT: preserve the channels!
+        # Otherwise the Dice Loss will do all kinds of shennanigans.
+        # Because it always interprets the first axis as channel,
+        # and treats it differently (sums over it independently).
+        # This will lead to a very large dice loss that dominates over everything else.
+        fg_input, fg_target = input_[:, 0:1], target[:, 0:1]
```

Comment:
This is crucial, if we don't keep the channels then the dice loss does wrong things.